packages/@react-aria/checkbox/src/useCheckboxGroup.ts

```diff
@@ -0,0 +1,62 @@
+/*
+ * Copyright 2020 Adobe. All rights reserved.
+ * This file is licensed to you under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License. You may obtain a copy
+ * of the License at http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under
+ * the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR REPRESENTATIONS
+ * OF ANY KIND, either express or implied. See the License for the specific language
+ * governing permissions and limitations under the License.
+ */
+
+import {AriaCheckboxGroupProps} from '@react-types/checkbox';
+import {filterDOMProps, mergeProps} from '@react-aria/utils';
+import {HTMLAttributes} from 'react';
+import {useLabel} from '@react-aria/label';
+
+interface CheckboxGroupAria {
+  /** Props for the checkbox group wrapper element. */
+  checkboxGroupProps: HTMLAttributes<HTMLElement>,
+  /** Props for the checkbox group's visible label (if any). */
+  labelProps: HTMLAttributes<HTMLElement>
+}
+
+/**
+ * Provides the behavior and accessibility implementation for a checkbox group component.
+ * Checkbox groups allow users to select multiple items from a list of options.
+ * @param props - Props for the checkbox group.
+ * @param state - State for the checkbox group, as returned by `useCheckboxGroupState`.
+ */
+export function useCheckboxGroup(props: AriaCheckboxGroupProps): CheckboxGroupAria {
+  let {
+    validationState,
+    isReadOnly,
+    isRequired,
+    isDisabled,
+    orientation = 'vertical'
+  } = props;
+
+  let {labelProps, fieldProps} = useLabel({
+    ...props,
+    // Checkbox group is not an HTML input element so it
+    // shouldn't be labeled by a <label> element.
+    labelElementType: 'span'
+  });
+
+  let domProps = filterDOMProps(props, {labelable: true});
+
+  return {
+    checkboxGroupProps: mergeProps(domProps, {
+      role: 'group',
+      'aria-invalid': validationState === 'invalid' || undefined,
+      'aria-errormessage': props['aria-errormessage'],
+      'aria-readonly': isReadOnly || undefined,
+      'aria-required': isRequired || undefined,
+      'aria-disabled': isDisabled || undefined,
+      'aria-orientation': orientation,
```

Comment:
@jnurthen do you know if there are any checkbox group patterns that would support this, similar to a radio group? Why does ARIA have one but not the other?

Reply:
Good catch, I forgot to recheck those. What should happen then? Should we maybe not provide those `aria-*` props here but pass those group props to the individual checkboxes? So one could disable the whole group instead of all checkboxes individually?

Reply:
Yeah I think that's probably the best option.

Reply:
Just want to reassure myself that we are on the same page - you are saying that the group should accept those props like `isReadOnly`, `isRequired`, `isDisabled` and "forward" them to individual checkbox, right?

I believe in such a case `isRequired` would have to be handled somehow differently? `isRequired` seems to mean that any of the checkboxes in a group has to be checked, not that all of them are checked. OTOH there are also other scenarios - like a list of agreements in which all of the checkboxes have to be checked, or a list of checkboxes with "select at least X". This would mean that this, in fact, up to the consumer to decide.

But as `aria-invalid` is not a valid attribute for a `role="group"` then there is no place to accept `validationState` right now as there is nothing that we could do automatically with it 🤔 Taking a look at `useRadioGroup` - `validationState` is part of the `AriaRadioGroupProps`, not `RadioGroupState`. So providing its value is always a responsibility of a consumer and this should be handled the same here.

Looking further at `useRadioGroup` - it accepts the very same props (`isReadOnly`, `isRequired`, `isDisabled`) but it just sets aria attributes based on them, it doesn't forward them anyhow to `useRadio` - even though it accepts them as well. So in the case of the radio group - the forwarding of those props has to be done by the consumer manually. This wiring happens within spectrum's `RadioGroup` & `Radio` components through context.

Reply:
So, thinking about this more, I think we should remove the `isRequired` and `validationState` from `CheckboxGroup`. `isRequired` is just passed through as the `required` DOM prop on each checkbox, which makes them all required as you said, not if some are required. AFAIK there's no way of doing that in HTML. We don't implement any validation logic ourselves, so this would be up to the app. Similar reasoning for `validationState`. Marking every checkbox as invalid doesn't really make sense either.

As for `isReadOnly` and `isDisabled`, I think propagating them to each of the children makes sense, along with `aria-disabled` on the group itself. RadioGroup does this through context, but one could do so in other ways too (e.g. cloneElement or something). I think we should leave this up to consumers of React Aria and not be too prescriptive about it.

Reply:
I've removed most of those props from this interface - the only one that I have left is `isDisabled` as `aria-disabled` is allowed on a `role="group"`.

`isRequired` was also left (but stays unused) because I extend `LabelableProps` and this interface supports `isRequired`:
https://github.com/adobe/react-spectrum/blob/8f7fbb33e34d228b3e2cb370634e4c2cc7ae9b0f/packages/%40react-types/shared/src/labelable.d.ts
although I'm not sure if this is not a mistake (I don't rly see it being used anywhere by `LabelableProps` consumers) and if it maybe shouldn't be moved to `SpectrumLabelableProps`. I've created a PR for such change if you agree with it: https://github.com/adobe/react-spectrum/pull/877